File: R/matrix.R
library(bench)
setwd("/Users/ancavanciupopescu/Desktop/Classes/CS 32/Final project/Data/")
A <- as.matrix(read.csv("A_data.csv")[,-1])
B <- as.matrix(read.csv("B_data.csv")[,-1])

# Matrix multiplication function
multiply_matrices <- function(A, B) {
  return(A %*% B)
}

# Matrix inversion function
invert_matrix <- function(A) {
  n <- dim(A)[1]
  regularization_term <- diag(1e-5, n, n)
  A_regularized <- A + regularization_term
  return(solve(A_regularized))
}

# Measure execution time and memory
results <- bench::mark(
  multiply = { mult_result <- multiply_matrices(A, B) },
  invert = { inv_result <- invert_matrix(A) },
  iterations = 10,
  check = FALSE
)
cat("Execution time for 500x500 matrix multiplication:", results$total_time[1], "\n")
cat("Memory allocation for 500x500 matrix multiplication:", results$mem_alloc[1]/1e6, "\n")
cat("Execution time from 500x500 matrix inversion:", results$total_time[2], "\n")
cat("Memory allocation from 500x500 matrix inversion:", results$mem_alloc[2]/1e6, "\n")

File: Python/matrix.py
import numpy as np
from memory_profiler import profile
import time
import pandas as pd

# Load data
path = "/Users/ancavanciupopescu/Desktop/Classes/CS 32/Final project/Data/A_data.csv"
A = pd.read_csv(path).iloc[:, 1:].values
path = "/Users/ancavanciupopescu/Desktop/Classes/CS 32/Final project/Data/B_data.csv"
B = pd.read_csv(path).iloc[:, 1:].values

# Matrix multiplication
@profile
def multiply_matrices(A, B):
    return np.dot(A, B)

# Matrix inversion
@profile
def invert_matrix(A):
    regularization_term = 1e-5 * np.eye(A.shape[0])
    A_regularized = A + regularization_term
    return np.linalg.inv(A_regularized)

# Measure execution and memory
start_time_mult = time.time()
result_mult = multiply_matrices(A, B)
end_time_mult = time.time()
mult_time = end_time_mult - start_time_mult
start_time_inv = time.time()
result_inv = invert_matrix(A)
end_time_inv = time.time()
inv_time = end_time_inv - start_time_inv
print("Execution time for 500x500 matrix multiplication", mult_time, "seconds")
print("Execution time for 500x500 matrix inversion:", inv_time, "seconds")